File: C++/代码随想录/Hashtable/202. Happy Number.cpp
// 202. Happy Number
// https://leetcode.com/problems/happy-number/

#include <iostream>

class Solution {
public:
    bool isHappy(int n) {
        int slow = n, fast = n;
        do {
            slow = getSquareSum(slow);
            fast = getSquareSum(getSquareSum(fast));
        } while (slow != fast);
        
        return slow == 1;
    }
    
    int getSquareSum(int n) {
        int sum = 0;
        while (n) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }
};

int main() {
    Solution sol;
    std::cout << sol.isHappy(19) << std::endl;
    
    return 0;
}

File: C++/代码随想录/Hashtable/202. Happy Number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def get_next(number):
            total_sum = 0
            while number > 0:
                number, digit = divmod(number, 10)
                total_sum += digit ** 2
            return total_sum

        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = get_next(n)

        return n == 1

sol = Solution()

# 测试用例
print(sol.isHappy(19))  # 应该输出 True，因为 1^2 + 9^2 = 82, 8^2 + 2^2 = 68, 6^2 + 8^2 = 100, 1^2 + 0^2 + 0^2 = 1
print(sol.isHappy(2))   # 应该输出 False，因为会进入一个不包含 1 的循环
